### PR TITLE
Fix infinite recursion in controlled gate definition

### LINKED ISSUE #4512: compare QuantumCircuit with open controlled gate raises RecursionError
### Information

- **Qiskit Terra version**: Master and 0.14.1
- **Python version**: 3.7
- **Operating system**: MacOS

### What is the current behavior?

```python
from qiskit import QuantumCircuit
from qiskit.circuit.library.standard_gates import XGate

circuit1 = QuantumCircuit(2)
circuit1.append(XGate().control(1, ctrl_state='0'), [0, 1])

circuit2 = QuantumCircuit(2)
circuit2.append(XGate().control(1, ctrl_state='0'), [0, 1])

circuit1 == circuit2
```
```
RecursionError: maximum recursion depth exceeded while calling a Python object
```

### What is the expected behavior?

These circuits should comparable and equal.

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

The controlled gate definition property was being evaluated and it was
an controlled X gate with an open control the generated definition
would reference itself because the working definition becomes defined
with self as an element in the list. That meant whenever the definition
was fully expanded ControlledGate.definition would infinitely recurse.
To workaround this issue this commit adds a shallow copy to break the
reference loop.

### Details and comments

Fixes #4512

### CODE REVIEW

File: test/python/circuit/test_controlled_gate.py [hunk 1]
```diff
@@ -781,6 +781,14 @@ def test_open_controlled_gate_raises(self):
         with self.assertRaises(CircuitError):
             base_gate.control(num_ctrl_qubits, ctrl_state='201')
 
+    def test_open_controlled_x_gate_eq(self):
+        """Test open control x gate equality."""
+        circuit1 = QuantumCircuit(2)
+        circuit1.append(XGate().control(1, ctrl_state='1'), [0, 1])
+        circuit2 = QuantumCircuit(2)
+        circuit2.append(XGate().control(1, ctrl_state='1'), [0, 1])
```
Comment (kdk) [inline test/python/circuit/test_controlled_gate.py:789]: I think these should be `ctrl_state='0'` to reproduce the original bug.

Comment (mtreinish) [reply test/python/circuit/test_controlled_gate.py:789]: Sigh, yeah I made a mistake when I was trying to get to the bottom of it but forgot to update the test. It turns out this isn't actually a fix the infinite recursion, it's still there. I'll push up a revert in a sec and I've reopened the original issue.

Comment (mtreinish) [reply test/python/circuit/test_controlled_gate.py:789]: Reverted in: https://github.com/Qiskit/qiskit-terra/pull/4516

**Comment by enavarro51** [comment]
@mtreinish and @1ucian0. I thought this was an interesting problem, so I spent some time debugging it and came up with what I think is a solution, but maybe not an elegant one.

The problem stems from the fact that there is no _define() method for CXGate, so calls to self.definition always go through the ControlledGate method definition(). For other x gates, like x or ccx, self.definition becomes a list that can be referenced.

With an open control on a cx, the definition() method creates a definition of [XGate, CXGate, XGate]. I believe the problem is that when the `__eq__` method in ControlledGate references self.definition, it ends up going through definition() multiple times. This seems to create the infinite regression.

I added a bit of a hack (amidst code that someone already referred to as 'hacky') to prevent the method from creating a new definition multiple times. This seems to work and passes all the tests. Here's the diff of what I did. 

Let me know if you would like me to do a PR. If there's a more elegant solution, I'd be very interested to see it. I can also add the open control test you referenced above.

```--- a/qiskit/circuit/controlledgate.py
+++ b/qiskit/circuit/controlledgate.py
@@ -100,6 +100,7 @@ class ControlledGate(Gate):
                     self.base_gate = base_gate
         self._ctrl_state = None
         self.ctrl_state = ctrl_state
+        self._open_cx = False
 
     @property
     def definition(self) -> List:
@@ -117,14 +118,17 @@ class ControlledGate(Gate):
         if self._definition:
             qreg = self._definition[0][1][0].register
             definition = self._definition
-        elif isinstance(self, CXGate):
+        elif isinstance(self, CXGate) and not self._open_cx:
             qreg = QuantumRegister(self.num_qubits, 'q')
             definition = [(self, [qreg[0], qreg[1]], [])]
         open_rules = []
-        for qind, val in enumerate(bit_ctrl_state[::-1]):
-            if val == '0':
-                open_rules.append([XGate(), [qreg[qind]], []])
+        if not isinstance(self, CXGate) or not self._open_cx:
+            for qind, val in enumerate(bit_ctrl_state[::-1]):
+                if val == '0':
+                    open_rules.append([XGate(), [qreg[qind]], []])
         if open_rules:
+            if isinstance(self, CXGate):
+                self._open_cx = True
             return open_rules + definition + open_rules
         else:
             return self._definition

```

**Comment by enavarro51** [comment]
@mtreinish Did this get resolved?

**Comment by kdk** [comment]
@enavarro51 This was fixed as part of #4444.

**Pull request merged**